ml/kmeans/SamplesBuffer.cc

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+//
+#include "SamplesBuffer.h"
+
+#include <fstream>
+#include <sstream>
+#include <string>
+
+void Sample::print(std::ostream &OS) const {
+    for (size_t Idx = 0; Idx != NumDims - 1; Idx++)
+        OS << CNs[Idx] << ", ";
+
+    OS << CNs[NumDims - 1];
+}
+
+void SamplesBuffer::print(std::ostream &OS) const {
+    for (size_t Idx = Preprocessed ? (DiffN + (SmoothN - 1) + (LagN)) : 0;
```

Comment:
Do values in for statement change per iteration? if not, is more efficient to calculate once. Does compiler take care of it?
Same question for some other for loops...

Reply:
Optimizing compilers should take care of them because they consist only of integral subexpressions.

Reply:
How about non-optimising compilers, can we make life easier to them, too?

Reply:
By definition we don't care about the performance of non-optimizing compilation modes. If your concern is legibility we can move the sum in a separate variable.

Reply:
I most welcome it :)